tests/e2e/v2tov3upgrade_test.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/cluster/gvk"
+	odhAnnotations "github.com/opendatahub-io/opendatahub-operator/v2/pkg/metadata/annotations"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/metadata/labels"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/utils/test/testf"
+
+	. "github.com/onsi/gomega"
+)
+
+const (
+	defaultCodeFlareComponentName = "default-codeflare"
+)
+
+type V2Tov3UpgradeTestCtx struct {
+	*TestContext
+}
+
+func v2Tov3UpgradeTestSuite(t *testing.T) {
+	t.Helper()
+
+	tc, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	// Create an instance of test context.
+	v2Tov3UpgradeTestCtx := V2Tov3UpgradeTestCtx{
+		TestContext: tc,
+	}
+
+	// Define test cases.
+	testCases := []TestCase{
+		{"codeflare present in the cluster before upgrade, after upgrade not removed", v2Tov3UpgradeTestCtx.ValidateCodeFlareSupportRemovalNotRemoveComponent},
+	}
+
+	// Run the test suite.
+	RunTestCases(t, testCases)
+}
+
+func (tc *V2Tov3UpgradeTestCtx) ValidateCodeFlareSupportRemovalNotRemoveComponent(t *testing.T) {
+	t.Helper()
+
+	nn := types.NamespacedName{
+		Name: defaultCodeFlareComponentName,
+	}
+
+	t.Cleanup(func() {
+		tc.DeleteResource(
+			WithMinimalObject(gvk.CodeFlare, nn),
+			WithIgnoreNotFound(true),
+			WithCustomErrorMsg("Failed to delete CodeFlare component resource '%s'", defaultCodeFlareComponentName),
+		)
+	})
+
+	dsc := tc.FetchDataScienceCluster()
+
+	dscOwnerReference := metav1.OwnerReference{
+		APIVersion:         gvk.DataScienceCluster.GroupVersion().String(),
+		Kind:               gvk.DataScienceCluster.Kind,
+		Name:               dsc.GetName(),
+		UID:                dsc.GetUID(),
+		BlockOwnerDeletion: ptr.To(true),
+		Controller:         ptr.To(true),
+	}
+	marshalledOwnerReference, err := json.Marshal(dscOwnerReference)
+	require.NoError(t, err)
+
+	tc.EnsureResourceCreatedOrPatched(
+		WithMinimalObject(gvk.CodeFlare, nn),
+		WithMutateFunc(testf.Transform(`
+		.metadata.ownerReferences = [%s] |
+		.metadata.labels["%s"] = "%s" |
+		.metadata.annotations["%s"] = "%s" |
+		.metadata.annotations["%s"] = "%s" |
+		.metadata.annotations["%s"] = "%s" |
+		.metadata.annotations["%s"] = "%s"`,
+			marshalledOwnerReference,
+			labels.PlatformPartOf,
+			strings.ToLower(gvk.DataScienceCluster.Kind),
+			odhAnnotations.PlatformVersion,
+			dsc.Status.Release.Version.String(),
+			odhAnnotations.PlatformType,
+			string(dsc.Status.Release.Name),
+			odhAnnotations.InstanceGeneration,
+			strconv.Itoa(int(dsc.GetGeneration())),
+			odhAnnotations.InstanceUID,
+			string(dsc.GetUID()),
+		)),
+		WithCustomErrorMsg("Failed to create or update CodeFlare component resource '%s'", defaultCodeFlareComponentName),
+	)
+
+	tc.triggerDSCReconciliation(t, "modelregistry", operatorv1.Managed)
+
+	tc.EnsureResourceExists(
```

Comment:
I think this should be followed or replaced by a `EnsureResourceExistsConsistently` since the cleanup logic could take a while and this could be be a false positive

Reply:
Updated 👍